[yargs] Updated definitions for @types/yargs

amikhalev opened the pull request:
Updated definitions for yargs 8.0 and fixed version number
Added tslint.json file and fixed all lint errors
Removed uses of Object and replaced with either any or a more specific
type
Fixed type of `choices` to accept undefined and true, because those are
accepted and functional values
Formatted everything

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 

```ts
// File: test.js
yargs.choices("i", [undefined, true, "value"]);
// All accepted
// node test.js
// node test.js -i
// node test.js -i value
```
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

poelstra commented:
Nice work!

Would it make sense to change `any` to `object` (where you changed it from `Object` to `any`, note the casing)?

(Also, would have been nice to have the formatting changes in a separate commit, but I suppose it's fairly readable still.)

paulvanbrenk commented:
The thing to remember is that `object` can not be `number`, where `any` can. Not sure that's a problem in this case or not.

pushplay commented:
Nice work @amikhalev.

I agree that small-o `object` is a better choice for these types.

I glanced through the yargs 7 release notes and didn't see any api changes that need to be accounted for.  v8 doesn't have any release notes, oddly.

amikhalev commented:
`object` does make sense for these types, I was mostly unsure of whether it was worth it to require `// TypeScript Version: 2.3` just for those. I think that the added safety probably is

amikhalev commented:
@poelstra @mizunashi-mana @pushplay @jeffkenney What do you think about specifying a type for the `argv` key other than `any`?

Something like
```ts
interface Args { // should probably come up with a better name
    _: string[];
    $0: string;
    [argName: string]: string | number | boolean | string[];
}

amikhalev commented:
Also, it looks like changing `Object` to `object` actually revealed a bug in the tests. Turns out that `t => t` is not actually  a valid `configParser` because it does not return a plain object.

pushplay commented:
> worth it to require // TypeScript Version: 2.3

I've never seen a policy explicitly spelled out anywhere, but I generally see declaration files unafraid of using the latest features.  Upgrading Typescript 2 point releases has generally been painless in my experience and should be encouraged.

> specifying a type for the argv key other than any

We could do, but you'd have to specify `[argName: string]: any` because the `option` `coerce` argument can transform an argument into another type.  This is useful because parse errors are reported neatly.  Here's an example where I coerce to a date:

```typescript
const argv = yargs
    .option("date", {
        alias: "d",
        coerce: (d: string) => {
            const date = new Date(d);
            if (isNaN(date.getTime())) {
                throw new Error(`Could not parse ${d} as a date.  Expected something like ${new Date().toISOString()}.`);
            }
            return date;
        },
        description: "date to start tailing from (default is to print the last block)",
        type: "string"
    })
```

> like changing Object to object actually revealed a bug in the tests

Nice!  That doesn't look like an example from the docs.  I believe I wrote that.

poelstra commented:
@amikhalev Good update, and nice catch, too :)
:+1: for the changes so far!

I agree the 'typed' argv is a good idea (even if the indexed parameters still need to be `any`). Gives a better sense of what it actually is.

leonard-thieu commented:
> object does make sense for these types, I was mostly unsure of whether it was worth it to require // TypeScript Version: 2.3 just for those. I think that the added safety probably is

`object` only requires 2.2 btw.

I got some bug reports when I bumped jQuery to TypeScript 2.3 (#17239). Using only 2.2 is probably fairly safe.

pushplay commented:
@leonard-thieu I feel your pain.  I had a similar hassle when updating the Elasticsearch declaration to a newer version of the API.

amikhalev commented:
> object only requires 2.2 btw.

Ah I see. Will change

I also found some keys that were added in 7.0.0 that are missing in `Options`, so I went ahead and added those.

Plus I fixed a few other signatures too
